code/datums/components/swabbing.dm

```diff
@@ -0,0 +1,114 @@
+/*!
+
+This component is used in vat growing to swab for microbiological samples which can then be mixed with reagents in a petridish to create a culture plate.
+
+*/
+/datum/component/swabbing
+	///The current datums on the swab
+	var/list/swabbed_items = list()
+	///Can we swab objs?
+	var/CanSwabObj
+	///Can we swab turfs?
+	var/CanSwabTurf
+	///Can we swab mobs?
+	var/CanSwabMob
+	///Callback for update_icon()
+	var/datum/callback/UpdateIcons
+	///Callback for update_overlays()
+	var/datum/callback/UpdateOverlays
+
+/datum/component/swabbing/Initialize(CanSwabObj = TRUE, CanSwabTurf = TRUE, CanSwabMob = FALSE, datum/callback/UpdateIcons, datum/callback/UpdateOverlays, swab_time = 10, max_items = 3)
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	RegisterSignal(parent, COMSIG_ITEM_PRE_ATTACK, .proc/TryToSwab)
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, .proc/examine)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS, .proc/handle_overlays)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_ICON, .proc/handle_icon)
+
+	src.CanSwabObj = CanSwabObj
+	src.CanSwabTurf = CanSwabTurf
+	src.CanSwabMob = CanSwabMob
+	src.UpdateIcons = UpdateIcons
+	src.UpdateOverlays = UpdateOverlays
+
+///Changes examine based on your sample
+/datum/component/swabbing/proc/examine(datum/source, mob/user, list/examine_list)
+	if(swabbed_items.len)
+		examine_list += "<span class='nicegreen'>There is a microbiological sample on [parent]!</span>"
+		examine_list += "<span class='notice'>You can see the following micro-organisms:</span>\n"
+		for(var/i in swabbed_items)
+			var/datum/biological_sample/samp = i
+			examine_list += samp.GetAllDetails(user.research_scanner) //Get just the names nicely parsed.
+
+///Ran when you attack an object, tries to get a swab of the object. if a swabbable surface is found it will run behavior and hopefully
+/datum/component/swabbing/proc/TryToSwab(datum/source, atom/target, mob/user, params)
+	set waitfor = FALSE //This prevents do_after() from making this proc not return it's value.
```

Comment:
This is bad.
Signals are called async, try to work around this fact to get a proper return.

Reply:
ninja told me to do this

Reply:
If he did he did, though you could also return the value and call the `do_after` async.

Reply:
sure

Reply:
this doesn't make any sense

Reply:
If you can't afford to sleep in this inner proc you should be using a timer with a callback to manage this